### PR TITLE
Add plugin to set recommended product dependencies

### PR DESCRIPTION
**Pull request by jared2501**
This change is [Reviewable](https://reviewable.io/reviews/palantir/sls-packaging/243)

### CODE REVIEW

**Comment by jared2501** [comment]
@jmcampanini @uschi2000

File: gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/RecommendedProductDependenciesExtension.groovy [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+
+import org.gradle.util.ConfigureUtil
+
+class RecommendedProductDependenciesExtension {
+
+    Set<RecommendedProductDependency> recommendedProductDependencies = []
+
+    void productDependency(Closure<?> closure) {
+        RecommendedProductDependency dep = new RecommendedProductDependency()
+        ConfigureUtil.configureUsing(closure).execute(dep)
```
Comment (jmcampanini) [inline gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/RecommendedProductDependenciesExtension.groovy:11]: this is cool!

File: gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/RecommendedProductDependenciesPlugin.groovy [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+import org.gradle.api.Project
+import org.gradle.api.plugins.JavaPlugin
+import org.gradle.api.tasks.bundling.Jar
+
+class RecommendedProductDependenciesPlugin implements Plugin<Project> {
+
+    @Override
+    void apply(Project project) {
+        project.plugins.apply('java')
+        def ext = project.extensions.create('recommendedProductDependencies', RecommendedProductDependenciesExtension)
```
Comment (jmcampanini) [inline gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/RecommendedProductDependenciesPlugin.groovy:14]: nit: name is long, but accurate. if there was a just as accurate, shorter name, would prefer that

File: gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/RecommendedProductDependenciesPlugin.groovy [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+
+        project.afterEvaluate {
+            ext.recommendedProductDependencies.each { recommendedProductDependency ->
+                def recommendedProductDeps = CreateManifestTask.jsonMapper.writeValueAsString(RecommendedProductDependencies.builder()
+                        .recommendedProductDependencies(ext.recommendedProductDependencies)
+                        .build())
+                Jar jar = (Jar) project.getTasks().getByName(JavaPlugin.JAR_TASK_NAME)
+                jar.manifest.attributes((CreateManifestTask.SLS_RECOMMENDED_PRODUCT_DEPS_KEY): recommendedProductDeps)
```
Comment (jmcampanini) [inline gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/RecommendedProductDependenciesPlugin.groovy:22]: whoa, seems rather straight forward. hard pressed to find an arguably better solution. this approach seems reasonable

File: gradle-sls-packaging/src/test/groovy/com/palantir/gradle/dist/RecommendedProductDependenciesPluginTest.groovy [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+        """.stripIndent()
+
+        when:
+        runSuccessfully(':jar')
+
+        then:
+        def jar = new File(projectDir, "build/libs/root-project.jar")
+        jar.exists()
+        def recommendedDeps = readRecommendedProductDeps(jar)
+        recommendedDeps == "{\"recommended-product-dependencies\":[{\"product-group\":\"group\",\"product-name\":\"name\",\"minimum-version\":\"1.0.0\",\"maximum-version\":\"1.x.x\",\"recommended-version\":\"1.2.3\"}]}"
```
Comment (jmcampanini) [inline gradle-sls-packaging/src/test/groovy/com/palantir/gradle/dist/RecommendedProductDependenciesPluginTest.groovy:34]: why not un-yaml and check params? isn't that what the consuming plugin will do?

**Review by jmcampanini — APPROVED** [review]
lgtm. just non-blocking comments

File: gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/RecommendedProductDependenciesPlugin.groovy (hunk 2 above)
Comment (jared2501) [reply gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/RecommendedProductDependenciesPlugin.groovy:14]: we couldn't come up with one

Comment (jared2501) [reply gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/RecommendedProductDependenciesPlugin.groovy:14]: ;(

File: gradle-sls-packaging/src/test/groovy/com/palantir/gradle/dist/RecommendedProductDependenciesPluginTest.groovy (hunk 4 above)
Comment (jared2501) [reply gradle-sls-packaging/src/test/groovy/com/palantir/gradle/dist/RecommendedProductDependenciesPluginTest.groovy:34]: talked offline and we thought this was good enough

**Pull request merged**